### ISSUE TITLE
Specs: HTML Template, cache and page life-cycle hooks 

### ISSUE THREAD

**Issue by frandiox** [OnsenUI/OnsenUI#2056]
Related issues: #1831 #1821 #1814 #1573
PRs: #2053

@anatoo @asial-matagawa @misterjunio @masahirotanaka 

### HTML Templates

All the platforms that we support already include `<template>` element so we should make Onsen UI compatible with it. We can leave `<ons-template>` as is for backward compatibility but prefer `<template>` element from now on. For older platforms, either Crosswalk or a simple [polyfill](https://github.com/webcomponents/template) enables `<template>` element.

```html
<template id="page.html">
  <ons-page>...</ons-page>
</template>
```

### Cache

Right now we are caching only "local" templates (included in `index.html`). We could also cache external files the first time Onsen UI reads them. Also, we can provide an optional `ons.preload([...]);` method that requests those templates and caches them. This would reduce the delay when pushing new pages located outside `index.html`.

### Page life-cycle hooks

Using `<template>` allows `<script>` tags inside it that run when the element is attached. This means that a page can be initialised like this:

```html
<template id="page1.html">
  <ons-page id="p1">
    Content

    <script>
      (function() {
         var page = document.querySelector('#p1');
         // initialise page
      })();
    </script>
  </ons-page>
</template>
```

This would require closures to avoid global pollution and could lead to timing issues with custom elements. Even though there are fixes for these issues, I think it might be better to just provide "life-cycle" hooks:

```html
<template id="page1.html">
  <ons-page id="p1">
    Content

    <script>
      document.querySelector('#p1').init = function() {
        // initialise page
      };
      // same for .show, .hide...
    </script>
  </ons-page>
</template>
```

These hooks will run at the same time as the events are fired.

For the naming, I think `init` could work, but if we want to be consistent with other components perhaps `onInit`, `onShow`, etc. are better (we already have `onAction`, `onInfiniteScroll`, etc.).

I started testing this in #2053.
Opinions?

**Comment by munsterlander** [comment]
I love most of everything but only have a few concerns.

- Regarding the template tag, my concern is that it would suggest Onsen can load any template and display it.  With the ons-template, you know it is Onsen only.  This may be a very small concern and with the overall user feedback, maybe it is what people want.
- Please add preload!!
- I know many users have requested the ability to run script tags, I am for one not sold on the idea.  I think life-cycle hooks are the better way of supporting this.  I personally do not like mixing my components, even though I do not adhere to SoC strictly.
- I do think standardizing event names to onInit, onShow is good.

All in all, keep up the amazing work!  I am sure I will have to relearn things, but progress is important to stay versatile.

**Comment by misterjunio** [comment]
Nothing to add, I think everything looks good 😃

**Comment by asial-matagawa** [comment]
@frandiox 
We can rewrite `text` of `script` element under environments where inline scripts (`<script>~~~~</script>`) are allowed.

For example:
```html
<DOCTYPE html>
<html>
<head>
    <script>
        class HogeComponent extends HTMLElement {

            // Can define constructor arguments if you wish.
            constructor() {
                // If you define a ctor, always call super() first!
                // This is specific to CE and required by the spec.
                super();

                console.log(`x-hoge#constructor`);
            }

            connectedCallback() {
                console.log(`x-hoge#connectedCallback`);
            }

            static get observedAttributes() {
                return ['foo'];
            }

            // Only called for the disabled and open attributes due to observedAttributes
            attributeChangedCallback(name, oldValue, newValue) {
                console.log(`x-hoge[${name}]: ${oldValue} -> ${newValue}`);
            }

            }

            class PiyoComponent extends HTMLElement {

            // Can define constructor arguments if you wish.
            constructor() {
                // If you define a ctor, always call super() first!
                // This is specific to CE and required by the spec.
                super();

                console.log(`x-piyo#constructor`);
            }

            connectedCallback() {
                console.log(`x-piyo#connectedCallback`);
            }

            static get observedAttributes() {
                return ['bar'];
            }

            // Only called for the disabled and open attributes due to observedAttributes
            attributeChangedCallback(name, oldValue, newValue) {
                console.log(`x-piyo[${name}]: ${oldValue} -> ${newValue}`);
            }

        }

        customElements.define('x-hoge', HogeComponent);
        customElements.define('x-piyo', PiyoComponent);
    </script>
    <template id="test">
        <script>
            console.log('in-template script 1');
        </script>

        <x-hoge foo="asdf">
            <x-piyo bar="qwer">
            <p></p>
            </x-piyo>
        </x-hoge>

        <script>
            console.log('in-template script 2');
        </script>
    </template>
</head>
<body>

    <script>
        console.log('Accessing content...');
        const content = document.getElementById('test').content;
        console.log('done.');

        console.log('Importing node tree...');
        const importedNode = document.importNode(content, true)
        const script = importedNode.querySelector('script');
        console.log('script: ' + script.text);
        script.text = `(function(){${script.text};console.log('injected');})();`;
        console.log('script: ' + script.text);
        console.log('done.');

        console.log('Appending child...');
        document.body.appendChild(importedNode);
        console.log('done.');
    </script>
</body>
</html>
```

Output:
```
Accessing content...
done.
Importing node tree...
x-hoge#constructor
x-hoge[foo]: null -> asdf
x-piyo#constructor
x-piyo[bar]: null -> qwer
script: 
            console.log('in-template script 1');
        
script: (function(){
            console.log('in-template script 1');
        ;alert('injected');})();
done.
Appending child...
in-template script 1
in-template script 2
x-hoge#connectedCallback
x-piyo#connectedCallback
done.
```

By using this, I think we can provide a simpler way to users.
At least we could remove `onInit` hook.

**Comment by frandiox** [comment]
@asial-matagawa We always limited the number of nodes in a template to 1 (only 1 `ons-page`), that's why I wrote the script tag inside `ons-page`, which works well. If we want to change this, it will also affect the old `ons-template` and could create issues.

If we keep the limitation of having only 1 element in templates, `ons-page` can deal with `script` tags easily: https://github.com/OnsenUI/OnsenUI/pull/2053/commits/91f792263a6fb46061d3e65b309312270455cfaf#diff-52c49156be6687c4b5dfefe178df4972R157
It can be removed during the constructor and attached later on. This was the first approach I tried but then I thought that using hooks would be better since we continue using the same concept (init, show, hide...). Also, it's probably safer since we make sure timing issues won't occur in any platform.

@anatoo @misterjunio @asial-matagawa What do you prefer, hooks or implementing some code to delay the `script` tags?

**Comment by asial-matagawa** [comment]
@frandiox 
>We always limited the number of nodes in a template to 1 (only 1 ons-page), that's why I wrote the script tag inside ons-page

Oh, sorry. I forgot about that.
I think
```
<template id="page1.html">
  <ons-page id="p1">
    Content

    <script>
      // code here
    </script>
  </ons-page>
</template>
```
is fine.

The events `onInit`, `onShow`, `onHide` and `onDestroy` are also great.
They can be usuful outside of `<template>` as well.

My suggestion is to run `<script>` at the time of `init` event without having users to be aware of `init` event by automatically replacing
```
<template id="page1.html">
  <ons-page id="p1">
    Content

    <script>
       // initialize page
    </script>
  </ons-page>
</template>
```
with
```
<template id="page1.html">
  <ons-page id="p1">
    Content

    <script>
      document.getElementById('p1').onInit = function() { // This line has been injected by Onsen UI

        // initialize page

      })(); // This line has been injected by Onsen UI
    </script>
  </ons-page>
</template>
```
since I guess most users will try to declare variables just under `<script>`, pollute global namespace and break something.
I think this would depend on your preference, so if you prefer no code injection, I will follow it.
Do you prefer this?

**Comment by asial-matagawa** [comment]
x: `try to declare variables just under <script>, pollute global namespace and break something.`
o: `try to access page contents just under <script> and break something.`

**Comment by frandiox** [comment]
@asial-matagawa I see what you mean now. I agree this is quite good for the user perspective, but the real difference is just 1 line (the wrapper). For this, we need to do some "magic" like modifying the script content and stuff, which means the feature depends more on Onsen UI core than on the browser. Also, I think it's a bit troublesome to get the current page because we cannot rely on an ID. We should perhaps add a hash attribute to the page or something like that.

Following that idea, maybe we can just inject a closure, bind "this" to the page and delay the script like in the commit I posted before. It would be nice to use: `<script>this.querySelector('ons-toolbar .center').innerHTML = 'Title';</script>`

However, I think I'd prefer to explicitly declare the hooks and avoid more magic (the hooks already bind "this" to the page). Also, in most of the cases, they can just use `myNavigator.topPage` to get the page: `<script>myNavigator.topPage.onInit = function() { this.querySelector(...).innerHTML = 'Title'; }</script>`

It's a tradeoff, let's discuss it @anatoo @misterjunio @asial-matagawa

**Comment by asial-matagawa** [comment]
@frandiox 
Thank you for clarifying!

>I think it's a bit troublesome to get the current page
Yes, maybe the implementation will be complex...

**Comment by misterjunio** [comment]
I agree with both approaches, on one hand usually 'magic' is easier for the user but yes, maybe it would be overly complicated and at the same time the hooks provide some more control, so I'd say we go for this simpler solution.

**Comment by asial-matagawa** [comment]
@misterjunio 
Actually both approaches can have the hooks.

(1) @frandiox 's suggestion
```html
<template id="page1.html">
  <ons-page>
    Content

    <script>
      myNavigator.topPage.onInit = function() {
        // Initialize page here
        var toolbarTitle = this.querySelector('ons-toolbar .center');

        toolbarTitle.innerHTML = 'Title';
      })();

      myNavigator.topPage.onShow = function() {
      })();

      myNavigator.topPage.onHide = function() {
      })();

      myNavigator.topPage.onDestroy = function() {
      })();
    </script>
  </ons-page>
</template>
```
I think this is similar to component definition in Vue 2.

(2) My suggestion
```html
<template id="page1.html">
  <ons-page>
    Content

    <script>
      // Initialize page here
      var toolbarTitle = this.querySelector('ons-toolbar .center');

      toolbarTitle.innerHTML = 'Title';

      this.onShow = function() {
      })();

      this.onHide = function() {
      })();

      this.onDestroy = function() {
      })();
    </script>
  </ons-page>
</template>
```
I think this is similar to controller definition in AngularJS 1.

**Comment by anatoo** [comment]
I prefer `(1) @frandiox 's suggestion`.

**Comment by masahirotanaka** [comment]
As the conclusion in the meeting, let's go with (1).

**Comment by asial-matagawa** [comment]
@frandiox @misterjunio 

@anatoo said that "It's weird if the execution context is automatically mutated." and @masahirotanaka said that "(1) is natural enough."

So, let's go with (1).
Thank you so much for the discussion!

**Comment by frandiox** [comment]
Thanks!

Therefore, this is all already implemented in #2053 :+1:
